### DISCUSSION TITLE
datasource golang-version provides non-downloadable version for .0 patch releases

### DISCUSSION THREAD

**Discussion by dornimaug** [renovatebot/renovate#20185] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.117.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

The datasource golang-version provides the version of go 1.20 as 1.20.0. The ".0" patch part is omitted in go releases at both the tags in https://github.com/golang/go as well as the download links for go on go.dev.

So this datasource cannot be used when the replaced version value is used to download packages from the original go.dev website for the first release of a new minor version. As I would think this is at least one of the intended use cases, I believe this is a bug and the version should be provided as "1.20".

Reproduction repository:
https://github.com/dornimaug/renovate-golang-version-zero-patch
Executing reproduce.sh on master works
Doing the same on the branch renovate/go-1.x does not

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
This datasource works for its intended current purpose, so this is not a bug.

Either:
- We could adapt it to your needs, if it doesn't break our current needs, or
- You could potentially use template `replace` capabilities to remove `\.0$` from the current versions

**Reply by dornimaug** [reply]
I might misunderstand the "current purpose" of this datasource, since it is not documented and - as I read it - not really explicitly stated in the original issue either https://github.com/renovatebot/renovate/issues/6798 You yourself commented there: https://github.com/renovatebot/renovate/issues/6798#issuecomment-997855762 "We also want to make sure that we don't propose upgrades to people before the downloads are ready." Which lead me to believe that it is indeed to be used for downloading golang from its website or that this would at least be one of the intended use cases. 
What is the purpose / what are your current needs?

I do use a replacement (implemented using "sed") in my actual project for now, but that feels like a workaround.

**Reply by dornimaug** [reply]
Btw. the script to install golang in the renovate docker images implements the same workaround:  https://github.com/containerbase/base/blob/2c00a4ed17268b05bf8137aae9c3e2f01cf6ba09/src/usr/local/buildpack/tools/v2/golang.sh#L39

**Reply by rarkins** [reply]
The point of my comment that you linked to is that if you propose an upgrade immediately after tags/releases appear on github, that doesn't mean it can be installed immediately. It's got nothing to do with major.minor

Current uses of the datasource can be found by searching the source code for golang-version: https://github.com/search?q=repo%3Arenovatebot%2Frenovate%20golang-version&type=code

Its values are sourced from: https://raw.githubusercontent.com/golang/website/HEAD/internal/history/release.go (you can see they have 3 part precision including for 1.20.0)

**Comment by dornimaug** [comment]
This is all the information I could gather so far:
1. golang uses 1.20 (omitting the patch version for the first release of a new minor version) as the version schema at least in the following cases: git-tags, website downloads, release notes, documentation and go.mod (where patch-versions are never specified), in fact in any place I could find, but one:
2. golang uses 1.20.0 as an additional tagging schema for their official images on dockerhub. I could not find any other uses of the 3-part scheme.
3. The renovate documentation does not specify the intent or use cases for the datasource. It is in fact empty: https://docs.renovatebot.com/modules/datasource/golang-version/
4. The original issue and PR do not specify the intent or use cases. At least not in such a clear manner that I would have thought that downloading packages from the website using the versions provided as is, was not intended.
5. Searching the code base reveals a single piece of information on the use in https://github.com/renovatebot/renovate/blob/719354793f970de9860a8b78ef34c793657fe3b4/lib/modules/versioning/go-mod-directive/readme.md about bumping the minor version in go.mod file, where only minor versions are ever specified, so would not be affected, if "1.20" is the value provided. Other uses in the code base include test cases which verify the go.mod-bumping or the inner workings of the golang-version, but I did not find any hints as to its intent in there.
6. After searching through all the non-archived repositories of "renovatebot" and "containerbase" I only found one other use of this datasource: the Dockerfile in renovate-docker-full, which uses it to update the argument of "install-tool" which - as already discussed - implements the same workaround as I did to strip the patch-part of the version if it is ".0". This would actually break if changed to 1.20.0, because install-tool verifies, that the argument is a semver.

To sum things up: go.mod never uses any patch-versions, which is the stated reason for the existence of the "go-mod-directive" versioning in renovate. If one wanted to keep the golang docker images up-to-date the "docker" datasource in combination with an appropriate versioning would seem to be the appropriate choice. The only place that requires a semver for reasons I could not find out is install-tool, which then strips the patch version part if it is "0" again when it actually uses the version to download the golang packages, which does not seem very clean.

I still genuinely do not understand what the intended use of the datasource golang-version is, that requires it to provide a version like 1.20.0. 
If it is only to satisfy the "adding a patch part to fake a semver - removing the patch part so it can actually be used"-logic in install-tool, is that really intentional and should not that be changed to use a schema that is more in line, with what the upstream provider uses and seems to be the actual purpose: downloading golang packages from go.dev/google?
Are there other intended purposes I still could not find after searching through code bases, documentation and issues?

**Reply by dornimaug** [reply]
Addendum: the internal go file used to source the version numbers may contain a zero for the Z field of the Version struct, but the authors of the package do explicitly state how the versioning scheme works, that is in contrast to semver and that trailing zeros are omitted for what they call "major releases" (1.20 would be a major release in go lingo): https://github.com/golang/website/blob/d7454c2ca8dd3b5b6623c63453f20450f4d2f72d/internal/history/history.go#L37
So the 0 value of the Z field is purely an internal implementation detail and also never used in the public API of said go package. The HTML data generated strips this part of the version number on purpose as documented.

**Comment by rarkins** [comment]
Renovate uses this datasource in two places, and it's also used in one external repo. If you were to modify the results so that trailing `.0` is stripped, you'd need to make sure that the two internal uses are adjusted if necessary, and also make sure that it's published in time for the v35 major release we are preparing, so that we can flag it as breaking to downstream users.

In other words, if you're interested in seeing the change, you can submit a PR.

**Reply by viceice** [reply]
we use the full version to be semver compatible. a change also needs a new golang-version versioning wich is based of semver but also coerces the short versions or treat them as range.

this would also need a change on containerbase golang installer to support the short versions.

**Reply by rarkins** [reply]
Perhaps [`semver-coerced`](https://github.com/renovatebot/renovate/tree/main/lib/modules/versioning/semver-coerced) would work?

**Reply by dornimaug** [reply]
@rarkins I am not sure which uses you are referring to. I only found one internal use where the `.0` actually matters here: https://github.com/renovatebot/docker-renovate-full/blob/fb81eee8077eb8f732ee5c24ce7f6cc9295980d0/Dockerfile#L42 
The parsing of the version in go.mod (for go-mod-directive) is already aware of the strictly two part schema in go.mod: https://github.com/renovatebot/renovate/blob/719354793f970de9860a8b78ef34c793657fe3b4/lib/modules/manager/gomod/extract.ts#L62
Can you please provide links to the the other internal and external repos you meant?

@viceice The golang installer itself already does the stripping, so if that was kept in place and the semver check, which happens somewhere higher up in the install-tool stack, was removed maybe this could be changed in a compatible manner which is able to deal with 1.20 and 1.20.0 as well.

**Comment by dornimaug** [comment]
As alternatives to changing the golang-version datasource and possibly creating a new versioning I see two more courses of action:
1. Add information to the docs of the golang-version datasource which states its purpose (which I am still a bit unsure about, so I would need your input to create such a PR) and informs users that they need to strip tailing `.0` parts to use the versions provided to download golang from its website. This might even be enough for now, because the workaround is trivial.
2. Wait for https://github.com/golang/go/issues/32450 to be resolved and Go itself to use semver which might happen years from now or even never.

**Reply by viceice** [reply]
maybe https://github.com/golang/go/issues/57631 comes earlier

**Reply by viceice** [reply]
so I would just document the behavior and hope go will do the step to add the zero sooner than later

**Comment by dornimaug** [comment]
This discussion and its underlying issue are now resolved, because Go started using 3-part versioning. 

From https://go.dev/doc/go1.21:
> Go 1.21 introduces a small change to the numbering of releases. In the past, we used Go 1.N to refer to both the overall Go language version and release family as well as the first release in that family. Starting in Go 1.21, the first release is now Go 1.N.0. Today we are releasing both the Go 1.21 language and its initial implementation, the Go 1.21.0 release. 

I can also confirm that renovate picked up the update and to make updates in my use case work, I had to *remove* the logic of stripping a `.0` suffix from the version-string prior to using it to download a package from go.dev
